Issue 6712: Table Segment index recovery tool

RaulGracia opened the pull request:
**Change log description**  
Adds a new Admin CLI command plus the associated instructions to repair a corrupted Table Segment index in storage from the main Table Segment data.

**Purpose of the change**  
Fixes #6712.

**What the code does**  
This PR tries to approach fixing #6712 from a repair perspective. In the original issue, we observed that the footer of the last index chunk for a Table Segment was corrupted. This could be due to a rare bug in Pravega, or even to a problem in the Storage system itself. In any case, we need tools to deal with such a situation in which a Table Segment is corrupted. To this end, this PR does the following:
- New Admin CLI `dataRecovery tableSegment-recovery`:
  - The assumption of this command is that the main Table Segment data (i.e., storage chunks where PUT/DELETE operations are stored) are available to reconstruct the index.
  - One input parameter for the command is the location (i.e., local folder) with the data chunks related to the Table Segment. This means that the user should be able to access this data. The second parameter is the name of the new Table Segment to be created to reconstruct the original one's index.
  - The command works in 3 main steps: i) ordered scan of the Table Segment data chunks to parse the `TableEntry` operations stored, ii) start a new local Pravega instance (with SLTS Storage enabled), iii) process all the Table Segment operations related to the parsed `TableEntry` and flush data to storage.
- With the above, the user should be able to have a local reconstruction of index for the original Table Segment data.
- After that, we also provide the instructions step-by-step on how to replace the newly reconstructed index chunks in the original cluster (container and storage metadata changes).

**How to verify it**  
New test added to test the new command.
Manual testing in progress.

RaulGracia commented:
@sachin-j-joshi @abhinb please, can you have a look?

abhinb commented:
At times we do see the below stacktrace generated while shutting down the in-process pravega which might confuse the user if the command has worked or not. 

We do see that the command could flush the output chunks.
`Flushed the Segment Container with containerId 0 to Storage.`

Post that is when we can see the below trace:-
> com.google.common.util.concurrent.ListenerCallQueue$PerListenerQueue run
SEVERE: Exception while executing callback: io.pravega.common.concurrent.Services$ShutdownListener@313c465f terminated({from = STOPPING})
java.util.concurrent.RejectedExecutionException: Task ThreadPoolScheduledExecutorService.ScheduledRunnable(id=1414392, isDelayed=false, scheduledTimeNanos=0, task=java.util.concurrent.Executors$RunnableAdapter@2e79c08[Wrapped task = io.pravega.segmentstore.server.reading.StreamSegmentReadIndex$$Lambda$2170/0x00000008008d4840@9253a1], future=java.util.concurrent.CompletableFuture@35ee855a[Not completed]) rejected from java.util.concurrent.ThreadPoolExecutor@f058c6b[Shutting down, pool size = 22, active threads = 1, queued tasks = 0, completed tasks = 1248059]
        at java.base/java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor.reject(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor.execute(Unknown Source)
        at io.pravega.common.concurrent.ThreadPoolScheduledExecutorService.execute(ThreadPoolScheduledExecutorService.java:243)
        at io.pravega.segmentstore.server.reading.StreamSegmentReadIndex.close(StreamSegmentReadIndex.java:175)
        at io.pravega.segmentstore.server.reading.ContainerReadIndex.closeIndex(ContainerReadIndex.java:425)
        at io.pravega.segmentstore.server.reading.ContainerReadIndex.lambda$closeAllIndices$0(ContainerReadIndex.java:434)
        at java.base/java.util.ArrayList.forEach(Unknown Source)
        at io.pravega.segmentstore.server.reading.ContainerReadIndex.closeAllIndices(ContainerReadIndex.java:434)
        at io.pravega.segmentstore.server.reading.ContainerReadIndex.close(ContainerReadIndex.java:115)
        at io.pravega.segmentstore.server.containers.StreamSegmentContainer.close(StreamSegmentContainer.java:270)
        at io.pravega.segmentstore.server.store.StreamSegmentContainerRegistry.unregisterContainer(StreamSegmentContainerRegistry.java:203)
        at io.pravega.segmentstore.server.store.StreamSegmentContainerRegistry.lambda$startContainerInternal$1(StreamSegmentContainerRegistry.java:186)
        at io.pravega.common.concurrent.Services$ShutdownListener.terminated(Services.java:126)
        at com.google.common.util.concurrent.AbstractService$3.call(AbstractService.java:95)
        at com.google.common.util.concurrent.AbstractService$3.call(AbstractService.java:92)
        at com.google.common.util.concurrent.ListenerCallQueue$PerListenerQueue.run(ListenerCallQueue.java:205)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
        at io.pravega.common.concurrent.ThreadPoolScheduledExecutorService$ScheduledRunnable.run(ThreadPoolScheduledExecutorService.java:193)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)

abhinb commented:
should the sizes of the rolled over chunks created by this command be the same as that on the affected cluster to maintain consistency with chunk sizes? In most cases the size is going to be the default 32M i suppose right?. But just in case it differs.

abhinb commented on cli/admin/src/main/java/io/pravega/cli/admin/dataRecovery/TableSegmentRecoveryCommand.java:
```diff
@@ -0,0 +1,282 @@
+                unprocessedBytesFromLastChunk = 0;
+                processedBytes += totalEntryLength;
+                // Show the progress of scanning Table Segment operations.
+                if (tableSegmentOperations.size() % 100 == 0) {
```
should we print the name of the chunk here. else it just looks like the percentage has fallen when we move to the next chunk file.

RaulGracia replied:
I have added the print file name before the execution of this method, so the progress related to that file.

RaulGracia commented:
@abhinb the message you posted may related to the shutdown of the local Pravega instance. I don't know if I can do much to do not show it. What I have added is a new print specifying that the command has been executed correctly.
I have also added the possibility to parameterize the rolling length of chunks.

sachin-j-joshi commented on cli/admin/src/main/java/io/pravega/cli/admin/dataRecovery/TableSegmentRecoveryCommand.java:
```diff
@@ -0,0 +1,288 @@
+        output("Storage directory for this cluster is: " + this.pravegaStorageDir);
+        FileSystemStorageConfig adapterConfig = FileSystemStorageConfig.builder()
+                .with(FileSystemStorageConfig.ROOT, this.pravegaStorageDir.toString())
+                .with(FileSystemStorageConfig.REPLACE_ENABLED, true)
```
REPLACE_ENABLED is not used in SLTS

RaulGracia replied:
Ok, removed.

sachin-j-joshi reviewed (commented):
One concern I have is that there are edge cases like during failover or zombie instances and other LTS failures. 

In such cases, we might end up writing extra junk data at the end of chunks or even extra junk chunks that should not be included in valid data.  This can also happen when there are zombie instances still writing to old chunks while at the same time new instance has started new chunks.

If we blindly enumerate all chunks on LTS and process then it might actually end up processing irrelevant/incorrect data.

Now, the SLTS segment/chunk metadata has correct start and end offsets for all valid data. 

One solution is to use this method to retrieve list of valid chunks directly from SLTS https://github.com/pravega/pravega/blob/master/segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/UtilsWrapper.java#L217

RaulGracia commented on cli/admin/src/main/java/io/pravega/cli/admin/dataRecovery/TableSegmentRecoveryCommand.java:
```diff
@@ -0,0 +1,288 @@
+                .with(FileSystemStorageConfig.ROOT, this.pravegaStorageDir.toString())
+                .with(FileSystemStorageConfig.REPLACE_ENABLED, true)
+                .build();
+        StorageFactory storageFactory = new FileSystemSimpleStorageFactory(ChunkedSegmentStorageConfig.DEFAULT_CONFIG, adapterConfig, this.executor);
```
I have reviewed the configuration. From the related parameters, I don't see anything dangerous from the default values for `maxIndexedSegments`, `maxIndexedChunksPerSegment`, `maxIndexedChunks`. In any case, if we ever hit any issue is just a config that can be change if ever needed.

RaulGracia commented:
Thanks @sachin-j-joshi, your comment about the source of the primary chunks for the tool may be relevant. However, this tool just assumes that the right chunks are provided as input, so the index can be reconstructed. How to correctly gather the primary data chunks is a different topic, so your observations would be important to be included in #6760. cc/ @abhinb
